include/revng/Model/Types.h

```diff
@@ -0,0 +1,903 @@
+
+struct TypesTree;
+
+/// \brief Enum for identifying different kinds of qualifiers.
+//
+// Notice that we are choosing to represent pointers and arrays as qualifiers.
+// The idea is that a qualifier is something that you can add to a type T to
+// obtain another type R, in such a way that if T is fully known also R is fully
+// known. In this sense Pointer and Array types are qualified types.
+namespace QualifierKind {
+enum Values { Invalid, Pointer, Array, Const };
+// TODO: fromString toString
+} // end namespace QualifierKind
+
+struct Qualifier;
+struct QualifiedType;
+
+struct EnumEntry;
+struct AggregateField;
+struct StructField;
+struct UnionField;
+struct ArgumentType;
+
+} // end namespace flatc
+
+// Make flatc::Type derived types usable with UpcastablePointer
+template<>
+struct concrete_types_traits<flatc::Type> {
+  using type = std::tuple<flatc::PrimitiveType,
+                          flatc::EnumType,
+                          flatc::TypedefType,
+                          flatc::StructType,
+                          flatc::UnionType,
+                          flatc::FunctionPtrType,
+                          flatc::Type>;
+};
+
+/// \brief Concept to identify all types that are derived from flatc::Type
+template<typename T>
+concept FlatCType = std::is_base_of_v<flatc::Type, T>;
+
+/// \brief Base class of flatc types used for LLVM-style RTTI
+struct flatc::Type {
+  static constexpr const char *Tag = "!flatc::InvalidType";
+
+  // TODO: we should switch to a sane UUID implementation
+  uint64_t UUID;
+
+  /// \brief The name of this type. This is the unique key to identifiy a type.
+  std::string Name;
+
+  /// \brief The kind of a flatc::Type
+  TypeKind::Values Kind;
+
+public:
+  TypeKind::Values getKind() const { return Kind; }
+
+  static bool classof(const Type *T) {
+    return T->getKind() == TypeKind::Invalid;
+  }
+
+  Type(llvm::StringRef NameRef, TypeKind::Values TK) :
+    UUID(0ULL), Name(NameRef), Kind(TK) {
+    auto Err = llvm::getRandomBytes(&UUID, sizeof(UUID));
+    if (Err)
+      UUID = 0ULL;
+  }
+
+  Type(llvm::StringRef NameRef) : Type(NameRef, TypeKind::Values::Invalid) {}
+  Type() : Type("") {}
+  ~Type() = default;
+
+  Type(const Type &) = default;
+  Type &operator=(const Type &) = default;
+
+  Type(Type &&) = default;
+  Type &operator=(Type &&) = default;
+
+  template<FlatCType T>
+  bool operator==(const T &Other) const {
+    return UUID == Other.UUID;
+  }
+
+public:
+  bool verify() const debug_function;
+
+private:
+  bool verifyBase() const debug_function;
+
+  std::strong_ordering compareBase(const Type &Other) const {
+    std::strong_ordering KindCmp = Kind <=> Other.Kind;
+    if (KindCmp != 0)
+      return KindCmp;
+
+    int NameCmp = llvm::StringRef(Name).compare(llvm::StringRef(Other.Name));
+    if (NameCmp < 0)
+      return std::strong_ordering::less;
+    if (NameCmp > 0)
+      return std::strong_ordering::greater;
+    return std::strong_ordering::equal;
+  }
```

Comment:
What's the purpose of this? It's not used anywhere.

Reply:
Leftover. Will remove.

Reply:
Done